troposphere/ec2.py

```diff
@@ -761,3 +762,58 @@ class VPCCidrBlock(AWSObject):
         'CidrBlock': (basestring, False),
         'VpcId': (basestring, True),
     }
+
+
+class TagSpecifications(AWSProperty):
+    props = {
+        'ResourceType': (basestring, False),
+        'Tags': ((Tags, list), False)
+    }
+
+
+class SpotOptions(AWSProperty):
+    props = {
+        'InstanceInterruptionBehavior': (basestring, False),
+        'MaxPrice': (basestring, False),
+        'SpotInstanceType': (basestring, False)
+    }
+
+
+class InstanceMarketOptions(AWSProperty):
+    props = {
+        'MarketType': (basestring, False),
+        'SpotOptions': (SpotOptions, False)
+    }
+
+
+class LaunchTemplateData(AWSProperty):
+    props = {
+        'BlockDeviceMappings': ([BlockDeviceMapping], False),
```

Comment:
So some of the properties you have in here share names with other EC2 properties, but are actually slightly different.  For example, this is the ec2 base BlockDeviceMapping:

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-blockdev-mapping.html

But what you actually want is the EC2 Launch Template Block Device Mapping:

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-blockdevicemapping.html

They look incredibly similar, but you'll notice that the `DeviceName` property of the latter is not required, while it is required on the former.

This makes me wonder if the docs are buggy, honestly.  In the end, we may want to name these differently.  Maybe `LaunchTemplateBlockDeviceMapping`?

There are a few of these, so you might want to go through those and verify as well (Placement and Monitoring seem like there might be collisions, I haven't gone through them all though).

As well, it might be worth investigating if the cloudformation schema file actually agrees with the docs.  Honestly, the LaunchTemplate docs in Cloudformation are kind of messy - like the ordering of the properties.

Reply:
I verified most of the class I reused or at least tried to.  Obvious had a miss there on BlockDeviceMappings.  I have done a really simple template that works.  Also discovered that CFN won't order it properly like a Launch Config and that you have to enforce using DependsOn. A reallllllly simple example.  I'll dig in more soon as well.  We are gonna get this figured out!

{
    "Description": "Configures autoscaling group for api app",
    "Resources": {
        "AutoscalingGroup": {
            "DependsOn": "LaunchTemplate",
            "Properties": {
                "DesiredCapacity": 1,
                "HealthCheckType": "EC2",
                "LaunchTemplate": {
                    "LaunchTemplateName": "test",
                    "Version": "1"
                },
                "MaxSize": 1,
                "MinSize": 1,
                "VPCZoneIdentifier": [
                    "subnet-71b9fd14"
                ]
            },
            "Type": "AWS::AutoScaling::AutoScalingGroup"
        },
        "LaunchTemplate": {
            "Properties": {
                "LaunchTemplateData": {
                    "ImageId": "ami-14c5486b",
                    "InstanceType": "c5.large"
                },
                "LaunchTemplateName": "test"
            },
            "Type": "AWS::EC2::LaunchTemplate"
        }
    }
}

Reply:
Fyi that was made with this tropo

from troposphere import Base64, Join
from troposphere import Parameter, Ref, Template
from troposphere.autoscaling import AutoScalingGroup, Tag
from troposphere.autoscaling import LaunchTemplateSpecification
import troposphere.ec2 as ec2

t = Template()

t.add_description("""\
Configures autoscaling group for api app""")

ltd = ec2.LaunchTemplateData(
    "LaunchTemplateData",
    ImageId='ami-14c5486b',
    InstanceType="c5.large"
)

LaunchTemplate = t.add_resource(ec2.LaunchTemplate(
    "LaunchTemplate",
    LaunchTemplateName='test',
    LaunchTemplateData=ltd
))

lts = LaunchTemplateSpecification("lts",
                                     LaunchTemplateName='test',
                                     Version="1")

AutoscalingGroup = t.add_resource(AutoScalingGroup(
    "AutoscalingGroup",
    DesiredCapacity=1,
    LaunchTemplate=lts,
    MinSize=1,
    MaxSize=1,
    VPCZoneIdentifier=['subnet-71b9fd14'],
    HealthCheckType="EC2",
    DependsOn='LaunchTemplate'
))

print(t.to_json())

Reply:
Man, the docs here just seem to be a big mess.  There are different types of BlockDeviceMapping Properties in the schema, though yeah, the reference doesn't seem to specify which type it means.  Maybe it's related to the current namespace?

```
    "AWS::EC2::Instance.BlockDeviceMapping": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-blockdev-mapping.html",
      "Properties": {
        "DeviceName": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-blockdev-mapping.html#cfn-ec2-blockdev-mapping-devicename",
          "PrimitiveType": "String",
          "Required": true,
          "UpdateType": "Mutable"
        },
        "Ebs": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-blockdev-mapping.html#cfn-ec2-blockdev-mapping-ebs",
          "Required": false,
          "Type": "Ebs",
          "UpdateType": "Mutable"
        },
        "NoDevice": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-blockdev-mapping.html#cfn-ec2-blockdev-mapping-nodevice",
          "Required": false,
          "Type": "NoDevice",
          "UpdateType": "Mutable"
        },
        "VirtualName": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-blockdev-mapping.html#cfn-ec2-blockdev-mapping-virtualname",
          "PrimitiveType": "String",
          "Required": false,
          "UpdateType": "Mutable"
        }
      }
    },
```

```
    "AWS::EC2::LaunchTemplate.BlockDeviceMapping": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-blockdevicemapping.html",
      "Properties": {
        "Ebs": {
          "Type": "Ebs",
          "Required": false,
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-blockdevicemapping.html#cfn-ec2-launchtemplate-blockdevicemapping-ebs",
          "UpdateType": "Mutable"
        },
        "NoDevice": {
          "Required": false,
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-blockdevicemapping.html#cfn-ec2-launchtemplate-blockdevicemapping-nodevice",
          "PrimitiveType": "String",
          "UpdateType": "Mutable"
        },
        "VirtualName": {
          "Required": false,
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-blockdevicemapping.html#cfn-ec2-launchtemplate-blockdevicemapping-virtualname",
          "PrimitiveType": "String",
          "UpdateType": "Mutable"
        },
        "DeviceName": {
          "Required": false,
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-blockdevicemapping.html#cfn-ec2-launchtemplate-blockdevicemapping-devicename",
          "PrimitiveType": "String",
          "UpdateType": "Mutable"
        }
      }
    },
```

And the DeviceName is required in some, but not in others.